ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -380,43 +350,66 @@ static void RefreshRootVnodeIfNecessary_Locked(VirtualizationRootHandle rootHand
 {
-    VirtualizationRootHandle rootIndex = FindUnusedIndexOrGrow_Locked();
-    
-    if (RootHandle_None != rootIndex)
+    VirtualizationRootHandle rootIndex;
+    do
     {
-        assert(rootIndex < s_maxVirtualizationRoots);
-        assert(!s_virtualizationRoots[rootIndex].inUse);
-        
-        // Retain a strong reference to the vnode if we have an active provider on it
-        if (userClient != nullptr)
+        rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, persistentIds);
+        if (rootIndex >= 0)
         {
-            errno_t error = vnode_get(vnode);
-            assertf(error == 0, "The calling code should already hold an iocount on vnode, so vnode_get should never fail. error = %d", error);
+            return rootIndex;
         }
         
-        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
-        
-        root->providerUserClient = userClient;
-        root->providerPid = clientPID;
-        root->inUse = true;
-
-        root->rootVNode = vnode;
-        root->rootVNodeVid = vid;
-        KextLog_File(vnode, "InsertVirtualizationRoot_Locked: virtualization root inserted at index %d: (path: \"%s\", fsid: 0x%x:%x, inode: 0x%llx) directory vnode %p:%u, user client PID %d, IOUC %p.",
-            rootIndex, path, persistentIds.fsid.val[0], persistentIds.fsid.val[1], persistentIds.inode, KextLog_Unslide(vnode), vid, clientPID, KextLog_Unslide(userClient));
-        
-        root->rootFsid = persistentIds.fsid;
-        root->rootInode = persistentIds.inode;
-
-        if (path != nullptr)
+        rootIndex = FindUnusedIndex_Locked();
+        if (rootIndex < 0)
         {
-            strlcpy(root->path, path, sizeof(root->path));
+            // No space, resize array
+            uint16_t newLength = MIN(s_maxVirtualizationRoots * 2u, INT16_MAX + 1u);
+            if (newLength <= s_maxVirtualizationRoots)
+            {
+                KextLog_Error("FindOrInsertVirtualizationRoot_LockedMayUnlock: growing virtualization root array for root at '%s' failed: out of valid indices, array has reached %u items\n", path, s_maxVirtualizationRoots);
+                return RootHandle_None;
+            }
+        
+            // Must drop lock to safely allocate memory with blocking alloc
+            RWLock_ReleaseExclusive(s_virtualizationRootsLock);
+            VirtualizationRoot* grownArray = Memory_AllocArray<VirtualizationRoot>(newLength);
+            KextLog_Info("FindOrInsertVirtualizationRoot_LockedMayUnlock: growing virtualization root array to %u items for root at '%s' with non-blocking alloc failed, blocking alloc returned %p\n", newLength, path, grownArray);
+            RWLock_AcquireExclusive(s_virtualizationRootsLock);
+            
+            uint16_t newLengthAgain = MIN(s_maxVirtualizationRoots * 2u, INT16_MAX + 1u);
+            if (newLengthAgain != newLength || newLength == s_maxVirtualizationRoots)
```

Comment:
Why do we need to check `newLength == s_maxVirtualizationRoots`?  

I'm having trouble finding a scenario where `newLengthAgain != newLength` would be false but `newLength == s_maxVirtualizationRoots` would be true.

Reply:
* `newLengthAgain != newLength` is of course `true` if another thread has beaten us to the resize.
 * However, if the other thread resizing the array caused it to hit the maximum of `INT16_MAX + 1u` items, then `newLength` and `newLengthAgain` will both be equal to that value, as will `s_maxVirtualizationRoots`. In that case, there's not really any "resizing" happening.

So I think the code is correct, but the intention is perhaps not clear. I think I might replace  `INT16_MAX + 1u` with a symbolic constant and compare `s_maxVirtualizationRoots` to that as the second part of this condition.

Reply:
> However, if the other thread resizing the array caused it to hit the maximum of INT16_MAX + 1u items, then newLength and newLengthAgain will both be equal to that value, as will s_maxVirtualizationRoots. In that case, there's not really any "resizing" happening.

Ok that makes sense thanks!

> I think I might replace INT16_MAX + 1u with a symbolic constant and compare s_maxVirtualizationRoots to that as the second part of this condition.

I think that would make this check much easier to follow, could you make that change before merging?

Reply:
> I think that would make this check much easier to follow, could you make that change before merging?

Thanks for reminding me - I've now tidied that up.